### PR TITLE
`yarn run` with no parameters prints the list of available commands

### PR DESCRIPTION
**Pull request by mollyIV**
### Checklist 

- [x] I've run `yarn test:unit` from the root directory to see all new and existing unit tests pass
- [x] I've run `yarn test:integration` from the root directory to see all new and existing integration tests pass
- [x] I've run `yarn test:lint` to ensure the code style is valid

### Description

> I'm a Yarn user who's just starting on a repository, and I'd like to have a better idea of the main scripts of the project.

closes https://github.com/yarnpkg/berry/issues/284

When a user runs `yarn run` command without any parameters, he / she will get the list of the available scripts in the project. 

I did manual testing including regression tests - calling `yarn run` with some parameters or skipping `run` keyword and calling the commands like `yarn test:unit` 👍 

Screen Shot 2019-07-09 at 18 30 15

### TODO / Questions 

* [x] Should we format the output differently, i.e. using bold font / color for script names? 🤔 
* [x] Would be great to add unit tests, any ideas how / if we could do that? 🤔 
* [x] `--json` support, which would be helpful for things like autocompletion - 👋 @arcanis, the idea is to format the output of the command `yarn run --json` in JSON format and print it to the console? 🤔 

Thanks you  🙇 🙇

### CODE REVIEW

**Comment by mollyIV** [comment]
Oh, `netlify/keen-leavitt-98f35d/deploy-preview — Deploy preview failed.` job has failed. Is it a known issue or something new due to the changes in this PR? 🤔

**Comment by arcanis** [comment]
>  Should we format the output differently, i.e. using bold font / color for script names? 🤔

Yep, I'd prefer to use a similar output to the one we use with `yarn config`!

>  Would be great to add unit tests, any ideas how / if we could do that? 🤔

We don't do unit tests for CLI commands, but we have integration tests that should be fairly simple to setup. Take a look at [`run.test.js`](https://github.com/yarnpkg/berry/blob/master/packages/acceptance-tests/pkg-tests-specs/sources/commands/run.test.js) to see how to write one.

>  --json support, which would be helpful for things like autocompletion - 👋 @arcanis, the idea is to format the output of the command yarn run --json in JSON format and print it to the console? 🤔

Even simpler! All you have to do is use `reportJson` and set the `json` flag in the reporter. Take a look at [`yarn config`](https://github.com/yarnpkg/berry/blob/master/packages/plugin-essentials/sources/commands/config.ts#L62) to get an idea.

> Is it a known issue or something new due to the changes in this PR?

It was a "no space left on device", so likely transient. I've restarted the build 🙂

File: packages/berry-cli/sources/cli.ts [hunk 1]
```diff
@@ -38,6 +39,25 @@ function runBinary(path: PortablePath) {
   }
 }
 
+async function printAvailableScripts(configuration: Configuration) {
+  const stdout = process.stdout;
```
Comment (arcanis) [inline packages/berry-cli/sources/cli.ts]: The CLI commands must us the `stdout` parameter passed in parameter to the action callback (see other commands).

File: packages/berry-cli/sources/cli.ts [hunk 2]
```diff
@@ -38,6 +39,25 @@ function runBinary(path: PortablePath) {
...
+  const report = await StreamReport.start({configuration, stdout}, async (report: StreamReport) => {
+    const {workspace} = await Project.find(configuration, configuration.startingCwd);
```
Comment (arcanis) [inline packages/berry-cli/sources/cli.ts]: The `Project.find` (and subsequent `WorkspaceRequiredError` check) should be done outside of the reporter.

File: packages/berry-cli/sources/cli.ts [hunk 3]
```diff
@@ -38,6 +39,25 @@ function runBinary(path: PortablePath) {
...
+      throw new WorkspaceRequiredError(configuration.startingCwd);
+
+
+    var projectCommands: string[] = []
+    workspace!.manifest.scripts.forEach((value: string, key: string) => {
+      projectCommands.push(`${' '}- ${key}`)
+      projectCommands.push(`  ${' '} ${value}`)
+    });
+
+    report.reportInfo(null, `Project commands\r\n${projectCommands.join('\r\n')}`);
```
Comment (arcanis) [inline packages/berry-cli/sources/cli.ts]: Prefer calling `reportInfo` once for each entry.

File: packages/berry-cli/sources/cli.ts [hunk 4]
```diff
@@ -67,9 +87,15 @@ async function run() {
       for (const command of plugin.commands || [])
         command(clipanion, pluginConfiguration);
 
-    clipanion.runExit(`yarn`, process.argv.slice(2), {
-      cwd: NodeFS.toPortablePath(process.cwd()),
-    });
+    // Print the list of available scripts if a `run` command is executed without the parameters
+    const noParamsPassedToRunCommand = process.argv.slice(2).length === 1 && process.argv[2] === `run`;
```
Comment (arcanis) [inline packages/berry-cli/sources/cli.ts]: This won't work - the command needs to be registered into our CLI framework 🙂 Basically, the idea is that you have to switch the `name` parameter from required (`<name>`) to optional (`[name]`) and then simply check whether it's undefined or not.

**Comment by mollyIV** [comment]
Thank you so much for doing a code review 🙇 

> the main one is that the logic must be added to the CLI framework rather than put into berry-cli (berry-cli is just a small launcher for the rest of the commands).

Just to be sure we are on the same page - it basically means I should've done the changes in [`run.ts`](https://github.com/yarnpkg/berry/blob/master/packages/plugin-essentials/sources/commands/run.ts), right? 🙂

File: packages/berry-cli/sources/cli.ts [hunk 5]
```diff
@@ -38,6 +39,25 @@ function runBinary(path: PortablePath) {
...
 
+async function printAvailableScripts(configuration: Configuration) {
+  const stdout = process.stdout;
+  const report = await StreamReport.start({configuration, stdout}, async (report: StreamReport) => {
+    const {workspace} = await Project.find(configuration, configuration.startingCwd);
+    if (!workspace)
+      throw new WorkspaceRequiredError(configuration.startingCwd);
+
+
+    var projectCommands: string[] = []
```
Comment (mollyIV) [inline packages/berry-cli/sources/cli.ts]: Just one more question: in yarn v1 we are also printing the commands available from binary scripts:

```
info Commands available from binary scripts: acorn, atob, babylon, browserslist, color-support, commitizen, detect-libc, errno, escodegen, esgenerate, eslint, esparse, esvalidate, flow, git-cz, git-release-notes, gulp, gunzip-maybe, handlebars, import-local-fixture, is-ci, jest, jest-runtime, js-yaml, jsesc, jsinspect, json5, loose-envify, miller-rabin, mkdirp, needle, node-pre-gyp, nopt, oc, opencollective, prettier, rc, regjsparser, rimraf, sane, semver, sha.js, shjs, sshpk-conv, sshpk-sign, sshpk-verify, strip-indent, strip-json-comments, uglifyjs, uuid, watch, webpack, which
```

Wondering if we should do the same in this task? 🤔

**Comment by arcanis** [comment]
> Just to be sure we are on the same page - it basically means I should've done the changes in run.ts, right? 🙂

Yep!

>

File: packages/berry-cli/sources/cli.ts (hunk 5 above)
Comment (arcanis) [reply packages/berry-cli/sources/cli.ts]: I'd put it behind a `[-v,--verbose]` option 🤔

File: packages/berry-cli/sources/cli.ts (hunk 4 above)
Comment (mollyIV) [reply packages/berry-cli/sources/cli.ts]: 👋 @arcanis , 

I have a problem with making a `name` an optional argument. I have been reading the documentation of [`clipanion`](https://github.com/arcanis/clipanion). I did the following changes in `run.ts` file:

```Typescript
.command(`run [name] [... args] [-T,--top-level]`)
```

Unfortunately I am getting the error:

```bash
Error: Expected "-", "--", "_", or [a-z] but "." found.
    at peg$buildStructuredError (/Users/mollyiv/Projects/berry/packages/berry-cli/bundles/berry.js:39091:12)
    at peg$parse (/Users/mollyiv/Projects/berry/packages/berry-cli/bundles/berry.js:40658:11)
    at Clipanion.command (/Users/mollyiv/Projects/berry/packages/berry-cli/bundles/berry.js:37332:26)
    at module.exports.exports.default (/Users/mollyiv/Projects/berry/packages/berry-cli/bundles/berry.js:121033:6)
    at run (/Users/mollyiv/Projects/berry/packages/berry-cli/bundles/berry.js:159:17)
```

It looks like there is a parsing error from `clipanion`. 

Is the command pattern syntax that I am using incorrect? 🤔

**Comment by mollyIV** [comment]
> Even simpler! All you have to do is use reportJson and set the json flag in the reporter. Take a look at yarn config to get an idea.

👍 Do we want to have the same JSON format as we do in v1 (without `Commands available from binary scripts` and `{"type":"error" ...`)? 🤔 

```
{"type":"info","data":"Commands available from binary scripts: acorn, atob, babylon, browserslist, color-support, commitizen, detect-libc, errno, escodegen, esgenerate, eslint, esparse, esvalidate, flow, git-cz, git-release-notes, gulp, gunzip-maybe, handlebars, import-local-fixture, is-ci, jest, jest-runtime, js-yaml, jsesc, jsinspect, json5, loose-envify, miller-rabin, mkdirp, needle, node-pre-gyp, nopt, oc, opencollective, prettier, rc, regjsparser, rimraf, sane, semver, sha.js, shjs, sshpk-conv, sshpk-sign, sshpk-verify, strip-indent, strip-json-comments, uglifyjs, uuid, watch, webpack, which"}
{"type":"info","data":"Project commands"}
{"type":"list","data":{"type":"possibleCommands","items":["build","build-bundle","build-chocolatey","build-deb","build-dist","build-win-installer","changelog","commit","dupe-check","lint","pkg-tests","prettier","release-branch","test","test-coverage","test-only","test-only-debug","watch"],"hints":{"build":"gulp build","build-bundle":"node ./scripts/build-webpack.js","build-chocolatey":"powershell ./scripts/build-chocolatey.ps1","build-deb":"./scripts/build-deb.sh","build-dist":"bash ./scripts/build-dist.sh","build-win-installer":"scripts\\build-windows-installer.bat","changelog":"git-release-notes $(git describe --tags --abbrev=0 $(git describe --tags --abbrev=0)^)..$(git describe --tags --abbrev=0) scripts/changelog.md","commit":"git-cz","dupe-check":"yarn jsinspect ./src","lint":"eslint . && flow check","pkg-tests":"yarn --cwd packages/pkg-tests jest yarn.test.js","prettier":"eslint src __tests__ --fix","release-branch":"./scripts/release-branch.sh","test":"yarn lint && yarn test-only","test-coverage":"node --max_old_space_size=4096 node_modules/jest/bin/jest.js --coverage --verbose","test-only":"node --max_old_space_size=4096 node_modules/jest/bin/jest.js --verbose","test-only-debug":"node --inspect-brk --max_old_space_size=4096 node_modules/jest/bin/jest.js --runInBand --verbose","watch":"gulp watch"}}}
{"type":"error","data":"No command specified."}
```

So basically just 👇 

```
{"type":"list","data":{"type":"possibleCommands","items":["build","build-bundle","build-chocolatey","build-deb","build-dist","build-win-installer","changelog","commit","dupe-check","lint","pkg-tests","prettier","release-branch","test","test-coverage","test-only","test-only-debug","watch"],"hints":{"build":"gulp build","build-bundle":"node ./scripts/build-webpack.js","build-chocolatey":"powershell ./scripts/build-chocolatey.ps1","build-deb":"./scripts/build-deb.sh","build-dist":"bash ./scripts/build-dist.sh","build-win-installer":"scripts\\build-windows-installer.bat","changelog":"git-release-notes $(git describe --tags --abbrev=0 $(git describe --tags --abbrev=0)^)..$(git describe --tags --abbrev=0) scripts/changelog.md","commit":"git-cz","dupe-check":"yarn jsinspect ./src","lint":"eslint . && flow check","pkg-tests":"yarn --cwd packages/pkg-tests jest yarn.test.js","prettier":"eslint src __tests__ --fix","release-branch":"./scripts/release-branch.sh","test":"yarn lint && yarn test-only","test-coverage":"node --max_old_space_size=4096 node_modules/jest/bin/jest.js --coverage --verbose","test-only":"node --max_old_space_size=4096 node_modules/jest/bin/jest.js --verbose","test-only-debug":"node --inspect-brk --max_old_space_size=4096 node_modules/jest/bin/jest.js --runInBand --verbose","watch":"gulp watch"}}}
```

**Comment by arcanis** [comment]
> Do we want to have the same JSON format as we do in v1

No, I don't think it's a good idea. The v1 json format was log-based, meaning that you had one entry for each printed output. It's not very useful, as you still need to parse the data somehow afterwards.

In the v2 we're using a data-based formatting, meaning that we print one line of data for each relevant item in what we're logging (check `yarn config --json` for an example).

File: packages/berry-cli/sources/cli.ts (hunk 4 above)
Comment (arcanis) [reply packages/berry-cli/sources/cli.ts]: Ah errata - clipanion doesn't currently support optional positional arguments + rest (it really should, but that'll be for another time). So what we should do is:

```
run [... args] [-T,--top-level]
```

And then on the JS side:

```js
if (args.length === 0)
  return printRunListing();

const [name, ...rest] = args;
```

Makes sense?

Comment (mollyIV) [reply packages/berry-cli/sources/cli.ts]: It does, thank you! 🙇 Now we can have an issue with `--json` flag:

```
.command(`run [... args] [-T,--top-level] [--json]`)
```

```
$ yarn run --json                               
Error: Couldn't find a script named "--json".

Usage: yarn run [... args] [-T,--top-level] [--json]
```

🤔

Comment (mollyIV) [reply packages/berry-cli/sources/cli.ts]: It does, thank you! 🙇 Now we can have an issue with `--json` flag:

```
.command(`run [... args] [-T,--top-level] [--json]`)
```

```
$ yarn run --json                               
Error: Couldn't find a script named "--json".

Usage: yarn run [... args] [-T,--top-level] [--json]
```

🤔

Comment (arcanis) [reply packages/berry-cli/sources/cli.ts]: I took a closer look at clipanion and it seems we won't be able to implement `--json` at the moment - there's an ambiguity on whether `--json` is meant to be part of the option or `args` (for example in `yarn run foo --json`) that it isn't able to solve at the moment. I'll implement backtracking later, but for now let's put `--json` aside 🙂

File: packages/berry-cli/sources/cli.ts (hunk 3 above)
Comment (mollyIV) [reply packages/berry-cli/sources/cli.ts]: I implemented the output format similar to `config` command 👍

File: packages/berry-cli/sources/cli.ts (hunk 4 above)
Comment (mollyIV) [reply packages/berry-cli/sources/cli.ts]: Very well, let's skip `--json` for now, thanks for your investigation 🙇

**Comment by mollyIV** [comment]
The new output format:

Screen Shot 2019-07-10 at 20 01 40

**Comment by mollyIV** [comment]
It looks like there are some integration tests that have failed. Sorry, on it 👀

**Comment by mollyIV** [comment]
I've been trying to figure out what is wrong with the failing tests. 

```
...

Summary of all failing tests

● Commands › workspace foreach › should run scripts in parallel but following the topological order
 when run with --parallel --topological
● Commands › workspace foreach › should prefix the output with run with --verbose
● Commands › workspace foreach › should never run the scripts on workspaces that match the --exclude list
● Commands › workspace foreach › should not fall into endless loop if foreach cmd is the same as lifecycle script name
...

Test Suites: 1 failed, 41 passed, 42 total
Tests:          4 failed, 1 skipped, 364 passed, 369 total
Snapshots: 3 failed, 179 passed, 182 total
Time:           236.524s
```

The root cause seems to be changing the `run` command interface:

```
.command(`run [... args] [-T,--top-level]`)
```

Was wondering if this ☝️ could impacted the way we process the parameters of `foreach` command:

https://github.com/yarnpkg/berry/blob/2e5360b94422244773fd468bd6dc1a8fab6cd8b2/packages/plugin-workspace-tools/sources/commands/foreach.ts#L105-L108

I was trying to debug a `foreach` command, however couldn't get any print to the console. I built the plugin by executing `yarn build:cli --plugin @berry/plugin-workspace-tools`, but no luck. 

@arcanis can I kindly ask you for a help / thoughts on that? 🙏 Thank you 🙇

**Comment by arcanis** [comment]
Yep, it was just a matter of changing the way to obtain the script name. Should be fixed!

For the record, running `yarn build:cli --plugin ...` didn't work because dynamic plugins can override the builtin plugins (which I think might be useful in the future if someone wants to fix a problem in the default Yarn installation without having to completely fork the project). So the local `workspace-tools` plugin definition was overriding the one contained in the bundle you were building.

The ideal workflow to test plugins is to build them independently: `yarn build:plugin-workspace-tools` 😊

**Comment by arcanis** [comment]
Everything's green, the code is great, good to go!

**Comment by mollyIV** [comment]
That's a great news 💪 Thank you so much for the review and your help 🙇 That was my first `berry` contribution and definitely not the last one 😉

**Pull request merged**